Document the paper system properties (#82)
PaperMC/docs: docs/paper/admin/reference/system-properties.md

---
slug: /reference/system-properties
---

# Paper System Properties 

> These system properties can be set when you start your server allowing for the configuration of various settings.

Reviewer: ```suggestion
These system properties can be set when you start your server allowing for the configuration of various advanced settings.
```

Reviewer: in general, maybe a bigger disclaimer about "danger ahead"?

Author: maybe time for a `:::danger` tag here :)

## How they work

System properties are set when you start your server. For example, if you are using a `.bat` or a `.sh` file to start your server, you can add the system properties to the file. For example:

```bash
java -Dpaper.log-level=FINE -jar paper.jar
```

Where a `-D` is used to set a system property, and the system property is `paper.log-level` with a value of `FINE`. Otherwise, just add them to the start command

## List of System Properties

#### paper.playerconnection.keepalive:

- **default**: `30`

> - **description**: Controls how long the player connection will wait before closing in seconds.

Reviewer 2: ```suggestion
- **description**: Controls how long the player connection will wait before closing when not receiving any keepalives, in seconds.
```

#### paper.ticklist-excessive-delay-threshold:

- **default**: `60`
- **description**: Sets the excessive tick delay to the specified number of seconds.

#### timings.bypassMax:

- **default**: `unset`
- **description**: Allows for bypassing the max number of frames to send to the Aikar's Timings API. Setting this will not permit bypassing the limit unless the API is configured to allow it.

#### LetMeReload:

- **default**: `unset`
- **description**: Overrides the need to send confirmation message when using the `/reload` command.

#### paper.disableChannelLimit:

- **default**: `unset`

> - **description**: Disables the channel limit for the server.

Reviewer: what channel? I assume its number of plugin messaging channels?

#### net.kyori.adventure.text.warnWhenLegacyFormattingDetected:

> - **default**: `true`
> - **description**: Enables or disables the warning when legacy formatting is detected.

Reviewer 2: This is [disabled by default](https://github.com/PaperMC/Paper/commit/ec90a78595af4cddb6e711ab71de4ee73c59ca49).

```suggestion
- **default**: `false`
- **description**: Enables or disables the warning when legacy formatting is detected in a chat component.
```

- **default**: `unset`
- **description**: Enables explicit flushing of the network channel.

#### Paper.enable-sync-chunk-writes:

- **default**: `unset`
- **description**: Syncs writes on each write call. This has a performance impact, particularly on hard drives.

#### paper.debug-sync-loads:

- **default**: `unset`
- **description**: Enables debug logging for sync chunk loads.

#### Paper.ignoreWorldDataVersion:

- **default**: `unset`
- **description**: Ignores the world data version when loading a world. This is not recommended and will likely cause issues.

#### Paper.allowAsyncChunksSingleCore:

- **default**: `unset`
- **description**: Allows async chunk loading on single core systems.

#### debug.entities:

- **default**: `unset`
- **description**: Enables debug log messages for entity information.

#### Paper.bypassHostCheck:

- **default**: `unset`
- **description**: Bypasses the host pattern matching attempt for the client when connecting to the server.

#### paper.ticklist-warn-on-excessive-delay:

- **default**: `unset`
- **description**: Enables the warning when a tick list is scheduled with an excessive delay.

#### debug.rewriteForIde:

- **default**: `unset`

> - **description**: Removes the NMS revision from the stack trace to allow for easier debugging in IDEs.

Reviewer: this isn't only the stack trace, right? this is way more, this did rewrite all plugin calls or something iirc?

Author: I'm not sure honestly, I'd need someone else's input here. It seems to be used in the Commodore class in CB

Reviewer 3: Yes it should rewrite plugins calls from
like `org/bukkit/craftbukkit/1_19_R1/....`  to `org/bukkit/craftbukkit/....`

```java
getAndRemove.put( "org/bukkit/".concat( "craftbukkit/" + NMS_REVISION_PACKAGE ), NMS_REVISION_PACKAGE );
```

#### convertLegacySigns:

- **default**: `unset`
- **description**: Converts legacy signs to the new format.

#### paper.maxCustomChannelName:

- **default**: `64`
- **description**: Sets the largest size that a plugin channel can be.

#### Paper.maxSignLength:

- **default**: `80`
- **description**: Sets the maximum line length for signs.

#### Paper.FilterThreshhold:

- **default**: `8192`
- **min**: `8192`
- **description**: Sets the maximum size of a packet to be filtered.

#### Paper.minPrecachedDatafixVersion:

- **default**: `Minecraft world version + 1`
- **description**: If you are expecting to convert a large number of chunks you might consider setting this to only convert from a point onwards.

#### paper.maxChunkThreads:

- **default**: `4`

> - **description**: Sets the maximum number of threads to use for chunk loading. You cannot use more than half of the available cores.

Reviewer 2: This flag no longer exists since the introduction of the new chunk system.

Author: > The ones removed in the last two versions I have kept

As it was a recent change I left them. Should I change this?

Reviewer 4: Might as well remove them already, yea

Reviewer 2: Ah, didn't see that. I don't think it makes sense to keep removed flags in this reference page though, since we have versioned docs.

#### Paper.WorkerThreadCount:

- **default**: `8` or `number of cpus - 2`. Whichever is lower
- **description**: Sets the number of worker threads to use for chunk loading.

#### Paper.excessiveTELimit:

- **default**: `750`
- **description**: Splits tile entities into multiple packets if there are more than this many.

#### paper.maxChunkThreads:

- **default**: `4` but is capped at `number of cpus / 2`

> - **description**: Sets the maximum number of threads to use for chunk loading.

Reviewer 2: ditto

---
slug: /reference/system-properties
---

# Paper System Properties 

These system properties can be set when you start your server allowing for the configuration of various settings.

:::danger Danger Ahead

Setting flags for the JVM can alter how it operates and the same goes for the Paper server.
If you are unsure about what a flag does, it is recommended that you **do not use it**.

:::

## How they work

System properties are set when you start your server. For example, if you are using a `.bat` or a `.sh` file to start your server, you can add the system properties to the file. For example:

```bash
java -Dpaper.log-level=FINE -jar paper.jar
```

> 

Reviewer 5: I'd note here that windows (powershell?) requires quoting anything with `.`s in it. So something like `java "-Dpaper.log-level=FINE"`. Please double-check this, I'm unable to confirm at this moment.

Author: Checked it, You are right, have fixed

Where a `-D` is used to set a system property, and the system property is `paper.log-level` with a value of `FINE`. Otherwise, just add them to the start command

## List of System Properties

#### paper.playerconnection.keepalive:

- **default**: `30`
- **description**: Controls how long the player connection will wait before closing when not receiving any keepalives, in seconds.

#### timings.bypassMax:

> - **default**: `unset`

Reviewer 5: For all of these unset ones - some of them need to be explicitly set to `true` and others just check for non-null. Maybe make note or standardize this? I'm not sure that it really matters all that much, though.

Author: Yeah, It wouldn't be too hard to document but most I saw just used `!= null` checks... lmk if this is wanted

Reviewer 5: taking a second look should really be fine to just say set it to `true` for everything - as that does work regardless.

---
slug: /reference/system-properties
---

# Paper System Properties 

These system properties can be set when you start your server allowing for the configuration of various settings.

:::danger Danger Ahead

> Setting flags for the JVM can alter how it operates and the same goes for the paper server.

Reviewer 4: ```suggestion
Setting flags for the JVM can alter how it operates and the same goes for the Paper server.
```

If you are unsure about what a flag does, it is recommended that you **do not use it**.

:::

## How they work

System properties are set when you start your server. For example, if you are using a `.bat` or a `.sh` file to start your server, you can add the system properties to the file. For example:

```bash
java -Dpaper.log-level=FINE -jar paper.jar
```

Where a `-D` is used to set a system property, and the system property is `paper.log-level` with a value of `FINE`. Otherwise, just add them to the start command

## List of System Properties

#### paper.playerconnection.keepalive:

- **default**: `30`
- **description**: Controls how long the player connection will wait before closing when not receiving any keepalives, in seconds.

#### paper.ticklist-excessive-delay-threshold:

- **default**: `60`

> - **description**: Sets the excessive tick delay to the specified number of seconds.

Reviewer 6: This doesn't really explain what excessive ticks mean. Would be nice to further explain here.

Author: Seems like this flag has been removed :P - I might just remove this from the docs

#### timings.bypassMax:

- **default**: `unset`

> - **description**: Allows for bypassing the max number of frames to send to the Aikar's Timings API. Setting this will not permit bypassing the limit unless the API is configured to allow it.

Reviewer 6: number of frames?

#### LetMeReload:

- **default**: `unset`

> - **description**: Overrides the need to send confirmation message when using the `/reload` command.

Reviewer 6: I would say this disables the reload confirmation message, override isn't the word you want to use here as we are just getting rid of it entirely.

#### paper.disableChannelLimit:

- **default**: `unset`
- **description**: Disables the plugin channel limit for the server. This will disable the limit of 128 plugin channels per player.

#### net.kyori.adventure.text.warnWhenLegacyFormattingDetected:

- **default**: `false`
- **description**: Enables or disables the warning when legacy formatting is detected in a chat component.

#### Paper.DisableClassPrioritization:

- **default**: `unset`
- **description**: Disables the class prioritization system - mostly an issue when failing to relocate or shade properly.

#### Paper.disableFlushConsolidate:

- **default**: `unset`
- **description**: Disables the netty flush consolidation system.

#### Paper.debugDynamicMissingKeys:

- **default**: `unset`

> - **description**: Enables debug logging for missing keys in NBT objects.

Reviewer 6: Can we see an example of a log for a missing key?

Author: I started a 1.12 server, and then loaded the world with 1.19. I might have mis-interpreted the problem this solves, but the DFU had no errors here so I struggled to find an error. I also couldn't find any use of this in discord or the GH issues either.

#### paper.debug-sync-loads:

- **default**: `unset`
- **description**: Enables debug logging for sync chunk loads.

#### Paper.ignoreWorldDataVersion:

- **default**: `unset`
- **description**: Ignores the world data version when loading a world. This is not recommended and will likely cause issues.

#### debug.entities:

- **default**: `unset`
- **description**: Enables debug log messages for entity information.

#### Paper.bypassHostCheck:

- **default**: `unset`
- **description**: Bypasses the host pattern matching attempt for the client when connecting to the server.

#### paper.ticklist-warn-on-excessive-delay:

- **default**: `unset`
- **description**: Enables the warning when a tick list is scheduled with an excessive delay.

#### debug.rewriteForIde:

- **default**: `unset`
- **description**: Removes the NMS revision from the stack trace to allow for easier debugging in IDEs. 
It also remaps plugin CB calls to remove the version information.

#### convertLegacySigns:

- **default**: `unset`
- **description**: Converts legacy signs to the new format.

#### paper.maxCustomChannelName:

- **default**: `64`

> - **description**: Sets the largest size that a plugin channel can be.

Reviewer 6: Sets the max channel name size?

#### Paper.maxSignLength:

- **default**: `80`
- **description**: Sets the maximum line length for signs.

#### Paper.FilterThreshhold:

- **default**: `8192`
- **min**: `8192`
- **description**: Sets the maximum size of a packet to be filtered.

#### Paper.minPrecachedDatafixVersion:

- **default**: `Minecraft world version + 1`
- **description**: If you are expecting to convert a large number of chunks you might consider setting this to only convert from a point onwards.

#### Paper.WorkerThreadCount:

- **default**: `8` or `number of cpus - 2`. Whichever is lower
- **description**: Sets the number of worker threads to use for chunk loading.

#### Paper.excessiveTELimit:

- **default**: `750`
- **description**: Splits tile entities into multiple packets if there are more than this many.

#### io.papermc.paper.suppress.sout.nags:

- **default**: `unset`
- **description**: Suppresses the nag message about using System.out/System.err in a plugin.

#### paper.strict-thread-checks:

- **default**: `unset`

> - **description**: Overrides the status of the AsyncCatcher so that it will always throw an exception if it is not on the main thread.

Reviewer 6: I still wouldn't use override here, more like set or something.

#### Paper.skipServerPropertiesComments:

- **default**: `unset`
- **description**: Skips the comments in the server.properties file.

#### Paper.debugInvalidSkullProfiles:

- **default**: `unset`

> - **description**: Enables debug logging for invalid skull profiles. It appends an exception to the log.

Reviewer 6: Instead of appending exceptions, say that it "This logs any invalid skulls in the world with location information to help properly track them down.
